### ISSUE TITLE
brew doctor reports "no formulae" for 3rd party taps

### ISSUE THREAD

**Issue by b-reich** [Homebrew/brew#16878]
### `brew doctor` output

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  packer
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.2.12-9-gc1aac26
ORIGIN: https://github.com/Homebrew/brew
HEAD: c1aac260ebbe9c97b161a6604c1fc9d3d2dd4239
Last commit: 87 minutes ago
Core tap HEAD: 9041e8ff42840e5c9c1655c40f579caf9411a1ad
Core tap last commit: 87 minutes ago
Core tap JSON: 11 Mar 18:55 UTC
Core cask tap HEAD: e0f0d1d785a0ab65b47c334ed3dcc39b687cff14
Core cask tap last commit: 11 minutes ago
Core cask tap JSON: 11 Mar 18:55 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim -f
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_NO_INSTALL_FROM_API: set
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.44.0 => /opt/homebrew/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.4-arm64
CLT: 15.3.0.0.1.1708646388
Xcode: N/A
Rosetta 2: false
```

### What were you trying to do (and why)?

Installing packer from hashicorp tap, because of recent license changes 

### What happened (include all command output)?

```
❯ brew install hashicorp/tap/packer
==> Tapping hashicorp/tap
Cloning into '/opt/homebrew/Library/Taps/hashicorp/homebrew-tap'...
remote: Enumerating objects: 4184, done.
remote: Counting objects: 100% (485/485), done.
remote: Compressing objects: 100% (133/133), done.
remote: Total 4184 (delta 405), reused 397 (delta 352), pack-reused 3699
Receiving objects: 100% (4184/4184), 763.30 KiB | 10.04 MiB/s, done.
Resolving deltas: 100% (2752/2752), done.
Tapped 2 casks and 28 formulae (88 files, 1MB).

❯ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  packer
```

### What did you expect to happen?

```
❯ brew doctor
Your system is ready to brew.
```
brew doctor shoulnt report an keg without an formulae

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install hashicorp/tap/packer && brew doctor
```

**Comment by b-reich** [comment]
sorry had to lie about `My "brew doctor output" above says Your system is ready to brew. and am still able to reproduce my issue.` but this is the problem 😅

**Comment by b-reich** [comment]
also expierence this issue with other 3rd party taps like goreleaser

**Comment by cho-m** [comment]
This happens as the formula loaders used are not capable of handling situation when formula with same name exists in both core and 3rd party tap.
https://github.com/Homebrew/brew/blob/2d774654e23ff00e61e697e467b7fcfb02e7baae/Library/Homebrew/diagnostic.rb#L840-L841

One possible quick fix might be #16879 which can try to fall back to loader that can handle this situation.

**Comment by cho-m** [comment]
Fix is merged. I tested a few simple cases (e.g. this scenario via API and non-API mode). Let us know if there are still issues or any new issues after changes.
